README.md
desiR
=====

This package is used to rank, select, and prioritise items based on
multiple characteristics or criteria. The functions have mainly been
used for high-dimensional biology experiments and therefore the ‘items’
are genes, proteins, or metabolites.

The methods are described in this
[publication](https://peerj.com/articles/1444/) ([PubMed
entry](http://www.ncbi.nlm.nih.gov/pubmed/26644980)) using a microarray
data set, and additional information is in the package vignette on
[CRAN](https://cran.r-project.org/web/packages/desiR/).


NEWS.md
# desiR 1.2.2
* Made compliant with updated knitr requirements.

# desiR 1.2.1

* Fixed bug in d.overall() function. Now handles missing values instead of giving an error message.

# desiR 1.2

* Added citation information.

* Added d.rank() function which ranks raw values then maps to a 0-1 scale.

# desiR 1.1

* updated des.line() function so that arguments passed to the desirability function must be given in the des.args=c() argument. The ellipses (...) is now used to pass arguments to the plot function (e.g. xlim, lwd, lty, col). Note that scripts from previous versions will fail and will need to be updated from:

des.line(x1, "d.high", cut1=10 cut2=11)

to

des.line(x1, "d.high", des.args=c(cut1=10, cut2=11))


* Provided GitHub url and location for bug reports.


# desiR 1.0

* First version


R/d.low.R
#' @title Low  values are desirable
#'
#' @description Maps a numeric variable to a 0-1 scale such that low values are
#' desirable.
#'
#' @details Values less than \code{cut1} will have a high desirability. Values
#' greater than \code{cut2} will have a low desirability. Values between
#' \code{cut1} and \code{cut2} will have intermediate values.
#'
#' @param x Vector of numeric or integer values.
#' @param cut1,cut2 Values of the original data that define where the
#' desirability function changes.
#' @param des.min,des.max Minimum and maximum desirability values. Defaults to
#' zero and one, respectively.
#' @param scale Controls how steeply the function increases or decreases.
#'
#' @return Numeric vector of desirability values.
#' @seealso \code{\link{d.high}},  \code{\link{d.4pl}}
#' 
#' @examples
#' set.seed(1)
#' x <- rnorm(1000, mean=100, sd =5) # generate data
#' d <- d.low(x, cut1=90, cut2=110, scale=1)
#' 
#' # plot data
#' hist(x, breaks=30)
#' # add line
#' des.line(x, "d.low", des.args=c(cut1=90, cut2=110, scale=1))
#'
#' hist(x, breaks=30)
#' des.line(x, "d.low", des.args=c(cut1=90, cut2=110, des.min=0.1,
#' des.max=0.95, scale=1.5))

d.low <- function(x, cut1, cut2,  des.min = 0, des.max = 1, scale = 1){
  
  if(cut1 >= cut2) stop("cut1 must be less than cut2\n")
  if(des.min < 0 | des.min > 1) stop("des.min must be between zero and one\n")
  if(des.max < 0 | des.max > 1) stop("des.max must be between zero and one\n")
  if(scale <= 0) stop("scale must be greater than zero\n")
  
  y <- rep(NA,length(x))
  y <- ((x - cut2)/(cut1 - cut2))^scale
  y[x < cut1] <- 1
  y[x > cut2] <- 0

  # rescale:  des.min to des.max
  y <- (y * (des.max - des.min))  + des.min

  return(y)
}


R/d.ends.R
#' @title Extreme (both high and low) values are desirable
#'
#' @description Maps a numeric variable to a 0-1 scale such that values at the
#' ends of the distribution are desirable.
#'
#' @details Values less than \code{cut1} and greater than \code{cut4} will have
#' a high desirability. Values between \code{cut2} and \code{cut3} will have a
#' low desirability. Values between \code{cut1} and \code{cut2} and between
#' \code{cut3} and \code{cut4} will have intermediate values. This function is
#' useful when the data represent differences between groups; for example, log2
#' fold-changes in gene expression. In this case, both high an low values are of
#' interest.
#'
#' @param x Vector of numeric or integer values.
#' @param cut1,cut2,cut3,cut4 Values of the original data that define where the
#' desirability function changes.
#' @param des.min,des.max Minimum and maximum desirability values. Defaults to
#' zero and one, respectively.
#' @param scale Controls how steeply the function increases or decreases.
#'
#' @return Numeric vector of desirability values.
#' @seealso \code{\link{d.central}}
#' @examples
#' set.seed(1)
#' x <- rnorm(1000, mean=100, sd =5) # generate data
#' d <- d.ends(x, cut1=90, cut2=95, cut3=105, cut4=110, scale=1)
#'
#' # plot data
#' hist(x, breaks=30)
#' # add line
#' des.line(x, "d.ends",  cut1=90, cut2=95, cut3=105, cut4=110, scale=1)
#'
#' hist(x, breaks=30)
#' des.line(x, "d.ends",  cut1=90, cut2=95, cut3=105, cut4=110,
#' des.min=0.1, des.max=0.95, scale=1.5)


d.ends <- function(x, cut1, cut2, cut3, cut4, des.min = 0,
                     des.max = 1, scale = 1){

  if(cut1 >= cut2) stop("cut1 must be less than cut2\n")
  if(cut2 >= cut3) stop("cut2 must be less than cut3\n")
  if(cut3 >= cut4) stop("cut3 must be less than cut4\n")
  if(des.min < 0 | des.min > 1) stop("des.min must be between zero and one\n")
  if(des.max < 0 | des.max > 1) stop("des.max must be between zero and one\n")
  if(scale <= 0) stop("scale must be greater than zero\n")
  
  y <- rep(NA, length(x))
  for (i in 1:length(x)){
    if (is.na(x[i])) next
    if (x[i] <= cut1 | x[i] >= cut4)  y[i] <- 1
    if (x[i] >= cut2 & x[i] <= cut3) y[i] <- 0
    if (x[i] > cut1 & x[i] < cut2) y[i] <- ((x[i] - cut2)/(cut1 - cut2))^scale
    if (x[i] > cut3 & x[i] < cut4) y[i] <- ((x[i] - cut3)/(cut4 - cut3))^scale
  }

  # rescale:  des.min to des.max
  y <- (y * (des.max - des.min))  + des.min
  
  return(y)
}


R/d.high.R
#' @title High values are desirable
#'
#' @description Maps a numeric variable to a 0-1 scale such that high values are
#' desirable.
#'
#' @details Values less than \code{cut1} will have a low desirability. Values
#' greater than \code{cut2} will have a high desirability. Values between
#' \code{cut1} and \code{cut2} will have intermediate values.
#'
#' @param x Vector of numeric or integer values.
#' @param cut1,cut2 Values of the original data that define where the
#' desirability function changes.
#' @param des.min,des.max Minimum and maximum desirability values. Defaults to
#' zero and one, respectively.
#' @param scale Controls how steeply the function increases or decreases.
#'
#' @return Numeric vector of desirability values.
#' @seealso \code{\link{d.low}},  \code{\link{d.4pl}}
#' 
#' @examples
#' set.seed(1)
#' x <- rnorm(1000, mean=100, sd =5) # generate data
#' d <- d.high(x, cut1=90, cut2=110, scale=1)
#'
#' # plot data
#' hist(x, breaks=30)
#' # add line
#' des.line(x, "d.high",  cut1=90, cut2=110, scale=1)
#'
#' hist(x, breaks=30)
#' des.line(x, "d.high", cut1=90, cut2=110, des.min=0.1,
#' des.max=0.95, scale=1.5)


d.high <- function(x, cut1, cut2, des.min = 0, des.max = 1, scale = 1){

  if(cut1 >= cut2) stop("cut1 must be less than cut2\n")
  if(des.min < 0 | des.min > 1) stop("des.min must be between zero and one\n")
  if(des.max < 0 | des.max > 1) stop("des.max must be between zero and one\n")
  if(scale <= 0) stop("scale must be greater than zero\n")
  
  
  y <- rep(NA,length(x))
  y <- ((x - cut1)/(cut2 - cut1))^scale
  y[x < cut1] <- 0
  y[x > cut2] <- 1

  # rescale:  des.min to des.max
  y <- (y * (des.max - des.min))  + des.min
  

  return(y)
}


R/d.rank.R
#' @title Converts values to ranks then to desirabilities
#'
#' @description Values are ranked from low to high or high to low, and
#' then the ranks are mapped to a 0-1 scale.
#'
#' @details If low values of a variable are desirable set the argument
#' low.to.high=TRUE, otherwise low.to.high=FALSE.
#'
#' If both high and low values are of interest, take the absolute
#' value of the variable and use low.to.high=FALSE. See the example below.
#'
#' This function is less flexible than the others but it can be used
#' to compare the results with rank aggregation methods.
#'
#' @param x Vector of numeric or integer values.
#'
#' @param low.to.high If TRUE, low ranks have high desirabilities; if
#' FALSE, high ranks have high desirabilities.
#'
#' @param ties Specifies how to deal with ties in the data. The value
#' is passed to the 'ties.method' argument of the rank()
#' function. Default is 'min'. See help(rank) for more information.
#'
#' @return Numeric vector of desirability values.
#' 
#' @examples
#' set.seed(1)
#' x1 <- rnorm(1000, mean=100, sd =5) # generate data
#' d <- d.rank(x1, low.to.high=TRUE)
#' 
#' # plot data
#' hist(x1, breaks=30)
#' # add line
#' des.line(x1, "d.rank", des.args=c(low.to.high=TRUE))
#'
#' x2 <- rnorm(1000, mean=0, sd =5) # positive and negative values
#' # could be mean differences or t-statistics
#' hist(abs(x2), breaks=30)
#' # add line
#' des.line(abs(x2), "d.rank", des.args=c(low.to.high=FALSE))

d.rank <- function(x, low.to.high=TRUE, ties="min"){

    if(is.logical(low.to.high) == FALSE) stop("low.to.high must be TRUE or FALSE\n")

    if (low.to.high) { # low ranks desirable
        
        y <- rank(x, ties.method=ties)
        # reverse ranking
        y <- length(y) + 1 - y
        
    } else { # high ranks desirable
        
        y <- rank(x, ties.method=ties)
    }
  
    # rescale to 0-1
    y <- (y-min(y))/(max(y)-min(y))

    return(y)
}


R/d.central.R
#' @title  Central values are desirable
#'
#' @description Maps a numeric variable to a 0-1 scale such that values in the
#' middle of the distribution are desirable.
#'
#' @details Values less than \code{cut1} and greater than \code{cut4} will have
#' a low desirability. Values between \code{cut2} and \code{cut3} will have a
#' high desirability. Values between \code{cut1} and \code{cut2} and between
#' \code{cut3} and \code{cut4} will have intermediate values. This function is
#' useful when extreme values are undesirable. For example, outliers or values
#' outside of allowable ranges. If \code{cut2} and \code{cut3} are close to each
#' other, this function can be used when a target value is desirable.
#'
#' @param x Vector of numeric or integer values.
#' @param cut1,cut2,cut3,cut4 Values of the original data that define where the
#' desirability function changes.
#' @param des.min,des.max Minimum and maximum desirability values, defaults to
#' zero and one, respectively.
#' @param scale Controls how steeply the function increases or decreases.
#'
#' @return Numeric vector of desirability values.
#' @seealso \code{\link{d.ends}}
#' 
#' @examples
#' set.seed(1)
#' x <- rnorm(1000, mean=100, sd =5) # generate data
#' d <- d.central(x, cut1=90, cut2=95, cut3=105, cut4=110, scale=1)
#'
#' # plot data
#' hist(x, breaks=30)
#' # add line
#' des.line(x, "d.central",  cut1=90, cut2=95, cut3=105, cut4=110, scale=1)
#'
#' hist(x, breaks=30)
#' des.line(x, "d.central",  cut1=90, cut2=95, cut3=105, cut4=110,
#' des.min=0.1, des.max=0.95, scale=1.5)
#'
#' # target value
#' hist(x, breaks=30)
#' des.line(x, "d.central", cut1=90, cut2=99.9, cut3=100.1, cut4=110)

d.central <- function(x, cut1, cut2, cut3, cut4, des.min = 0,
                     des.max = 1, scale = 1){

  if(cut1 >= cut2) stop("cut1 must be less than cut2\n")
  if(cut2 >= cut3) stop("cut2 must be less than cut3\n")
  if(cut3 >= cut4) stop("cut3 must be less than cut4\n")
  if(des.min < 0 | des.min > 1) stop("des.min must be between zero and one\n")
  if(des.max < 0 | des.max > 1) stop("des.max must be between zero and one\n")
  if(scale <= 0) stop("scale must be greater than zero\n")

  
  y <- rep(NA, length(x))
  for (i in 1:length(x)){
    if (is.na(x[i])) next
    if (x[i] <= cut1 | x[i] >= cut4)  y[i] <- 0
    if (x[i] >= cut2 & x[i] <= cut3) y[i] <- 1
    if (x[i] > cut1 & x[i] < cut2) y[i] <- ((x[i] - cut1)/(cut2 - cut1))^scale
    if (x[i] > cut3 & x[i] < cut4) y[i] <- ((x[i] - cut4)/(cut3 - cut4))^scale
  }

  # rescale:  des.min to des.max
  y <- (y * (des.max - des.min))  + des.min
  
  return(y)
}
  


R/d.overall.R
#' @title Combine individual desirabilities
#'
#' @description Combines any number of desirability values into an overall
#' desirability.
#'
#' @details This function takes any number of individual desirabilities and
#' combines them with a weighted geometric mean to give an overall
#' desirability. The weights should be chosen to reflect the importance of the
#' variables. The values of the weights do not matter, only their relative
#' differences. Therefore weights of 4, 2, 1 are the same as 1, 0.5, 0.25. In
#' both cases the second weight is half of the first, and the third weight is a
#' quarter of the first.
#'
#' @param ... Any number of individual desirabilities.
#' @param weights Allows some desirabilities to count for more in the overall
#' calculation. Defaults to equal weighting.
#'
#' @return Numeric vector of desirability values.
#' 
#' @examples
#' set.seed(1)
#' x1 <- rnorm(1000, mean=100, sd =5) # generate data
#' x2 <- rnorm(1000, mean=100, sd =5) 
#' 
#' d1 <- d.high(x1, cut1=90, cut2=110, scale=1)
#' d2 <- d.low(x2, cut1=90, cut2=110, scale=1)
#'
#' D <- d.overall(d1, d2, weights=c(1, 0.5))
#' plot(rev(sort(D)), type="l")

d.overall <- function(..., weights = NULL){
  
  # get number of variables
  n <- length(list(...))

  # check whether any variables passed
  if(n == 0) stop("Some desirabilities must be included\n")

  # merge into matrix
  d.all <- cbind(...)

  if(min(d.all, na.rm=TRUE) < 0 | max(d.all, na.rm=TRUE) > 1) {
      stop("Desirabilities must be between 0 and 1\n")
  }

  # check lengths match
  if(!is.null(weights)) {
    if(ncol(d.all) != length(weights) ) {
      stop("Number of weights does not match number of desirabilities\n")
    }
  }
  
  # equal weights if none provided
  if(is.null(weights) == TRUE) {
    w <- rep(1,n)/n
  } else {
    w <- weights
  }
  
  # weighted geometric mean by row (omit missing)
  y <- apply(d.all, 1, function(x) exp(sum(w[!is.na(x)] * log(x[!is.na(x)])/
                                           sum(w[!is.na(x)])) )
             )
  
  return(y)
}


R/des.line.R
#' @title Plots a desirability function on an existing graph
#'
#' @description Plots any of the desirability functions on top of a graph,
#' usually a histogram or density plot.
#'
#' @details This function can be used to visualise how the desirabilities are
#' mapped from the raw data to a 0-1 scale, which can help select suitable cut
#' points. The scale of the y-axis has a minimum of 0 and a maximum of 1.
#'
#' WARNING: If you set xlim values for the histogram or density plot, then you must pass the same xlim values to des.line; otherwise the data and desirability function (plotted line) will be misaligned. If xlim is not set, then the same default values will be used for the data and the function. 
#'
#' @param x Vector of numeric or integer values.
#' @param des.func Name of the desirability function to plot (in quotes).
#' @param des.args A vector of named arguments for the chosen desirability
#' function.
#' @param ... Arguments for the plotting function (e.g. xlim, lwd, lty).
#'
#' @return Plotted values of the desirability function.
#' @seealso \code{\link{d.low}}, \code{\link{d.high}}, \code{\link{d.central}},
#' \code{\link{d.ends}}, \code{\link{d.4pl}}
#' 
#' @examples
#' set.seed(1)
#' x1 <- rnorm(100, 10, 2)
#' hist(x1, breaks=10, col="grey", border="white")
#' des.line(x1, "d.high", des.args=c(cut1=10, cut2=11))
#' des.line(x1, "d.high", des.args=c(cut1=10, cut2=11,
#' des.min=0.1, scale=0.5))



des.line <- function(x, des.func, des.args, ...){

  # check function selected correctly (one must be specified)
  if (!des.func %in% c("d.low","d.high","d.central", "d.ends", "d.4pl")) {
    stop("\ndes.func must be one of 'd.high', 'd.low',
         'd.central', 'd.ends', or 'd.4pl'\n")
  }

  # make values for plotting
  vals <- seq(min(x, na.rm=TRUE), max(x, na.rm=TRUE), length.out=300)

  # add lines to existing graph
  graphics::par(new=TRUE)
  
  if (des.func=="d.high") {
   graphics:: plot(do.call(d.high, c(list(x=vals), des.args) ) ~ vals,
         ylim=c(0,1), type="l",  yaxt="n", bty="n", xaxt="n",
         ylab="", xlab="", ...)
  }

  
  if (des.func=="d.low") {
    graphics::plot(do.call(d.low, c(list(x=vals), des.args) ) ~ vals,
         ylim=c(0,1), type="l",  yaxt="n", bty="n", xaxt="n",
         ylab="", xlab="", ...)
  }

  
  if (des.func=="d.central") {
    graphics::plot(do.call(d.central, c(list(x=vals), des.args) ) ~ vals,
         ylim=c(0,1), type="l",  yaxt="n", bty="n", xaxt="n",
         ylab="", xlab="", ...)
  }

  
  if (des.func=="d.ends") {
    graphics::plot(do.call(d.ends, c(list(x=vals), des.args) ) ~ vals,
         ylim=c(0,1), type="l",  yaxt="n", bty="n", xaxt="n",
         ylab="", xlab="", ...)
  }

    
  if (des.func=="d.4pl") {
    graphics::plot(do.call(d.4pl, c(list(x=vals), des.args) ) ~ vals,
         ylim=c(0,1), type="l",  yaxt="n", bty="n", xaxt="n",
         ylab="", xlab="", ...)
  }

}
